crates/sui-core/src/streamer.rs

```diff
@@ -1,21 +1,83 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use sui_types::event::EventEnvelope;
-use tokio::{sync::mpsc::Receiver, task::JoinHandle};
-use tracing::debug;
 
-pub struct Streamer {
-    event_queue: Receiver<EventEnvelope>,
+use crate::event_filter::Filter;
+use crate::event_handler::EVENT_DISPATCH_BUFFER_SIZE;
+use futures::Stream;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use sui_types::base_types::ObjectID;
+use sui_types::error::SuiError;
+use tokio::runtime::Handle;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::{mpsc, RwLock};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{debug, warn};
+
+type Subscribers<T, F> = Arc<RwLock<BTreeMap<String, (Sender<T>, F)>>>;
+pub struct Streamer<T, F: Filter<T>> {
+    streamer_queue: Sender<T>,
+    subscribers: Subscribers<T, F>,
 }
 
-impl Streamer {
-    pub fn spawn(rx: Receiver<EventEnvelope>) -> JoinHandle<()> {
-        tokio::spawn(async move { Self { event_queue: rx }.stream().await })
+impl<T, F> Streamer<T, F>
+where
+    T: Clone + Debug + Send + Sync + 'static,
+    F: Filter<T> + Ord + Clone + Send + Sync + 'static + Clone,
+{
+    pub fn spawn(buffer: usize) -> Self {
+        let (tx, rx) = mpsc::channel::<T>(buffer);
+        let streamer = Self {
+            streamer_queue: tx,
+            subscribers: Default::default(),
+        };
+        let mut rx = rx;
+        let subscribers = streamer.subscribers.clone();
+        tokio::spawn(async move {
+            while let Some(data) = rx.recv().await {
+                Self::send_to_all_subscribers(subscribers.clone(), data).await;
+            }
+        });
+        streamer
     }
 
-    pub async fn stream(&mut self) {
-        while let Some(event_envelope) = self.event_queue.recv().await {
-            debug!(event_envelope =? event_envelope, "Get event");
+    async fn send_to_all_subscribers(subscribers: Subscribers<T, F>, data: T) {
+        for (id, (subscriber, filter)) in subscribers.read().await.clone() {
+            if !(filter.matches(&data)) {
+                continue;
+            }
+            let data = data.clone();
```

Comment:
is it possible to avoid clone here?  the old `data` does not seem used afterwards

Reply:
this is in a loop, so clone is needed for each subscriber